### PR TITLE
Allow SAML2 configuration to not assume file system

### PR DESCRIPTION
**Pull request by mmoayyed**
## Goals

Allow the metadata generation and keystore generation for pac4j-saml, acting as a service provider to not always assume a file-system. The client application may want to manage metadata, keystore, etc inside a separate external system (i.e. database). By removing the file-system assumption, we are now exposing options so that client applications can build a modest REST API to respond to pac4j requests for storing and fetching of SAML metadata, keystore, etc.

## How

- For metadata generation, we now have:

1. `SAML2FileSystemMetadataGenerator`: same as before; this is the existing default behavior.
2. `SAML2HttpUrlMetadataGenerator`: able to resolve/fetch and store metadata using a REST API that responds to GET/POST methods. This is primarily handled using `HTTPMetadataResolver` component of OpenSAML v4. The metadata content is always generated by pac4j-saml and then passed along to the API for storage. The API simply has to retrieve that metadata, or store it. No other extra processing/work is necessary.

- For keystore generation, we now have:

1. `SAML2FileSystemKeystoreGenerator`: same as before; this is the existing default behavior.
2. `SAML2HttpUrlKeystoreGenerator`: able to resolve/fetch and store keystore using a REST API that responds to GET/POST methods. The metadata content is always generated by pac4j-saml and then passed along to the API for storage as a base64 encoded string. The API simply has to retrieve that keystore, or store it. No other extra processing/work is necessary. Pac4j will decode it keystore back, construct it again and will proceed as normal.

## Notes

- Keeping separate APIs/components for metadata and keystore generation allows for maximum flexibility while maintaining backwards compatibility to the maximum.
- Additional tests are added to make sure all new components work as expected. All existing tests are kept as they were to pass.
- In practice, it's likely that `SAML2HttpUrlMetadataGenerator` would be used with `SAML2HttpUrlKeystoreGenerator`. However, given the flexibility that exists with separation of components, this is not strictly necessary.
- Contacting URL endpoints is done using `SAML2HttpClientBuilder` that builds on of OpenSAML v4 with minor adjustments. The builder can support basic-auth, r other forms of security for endpoint access.

## Example Config

```java
        SAML2HttpClientBuilder httpClient = new SAML2HttpClientBuilder();
        httpClient.setConnectionTimeout(Duration.ofSeconds(1));
        httpClient.setSocketTimeout(Duration.ofSeconds(1));

        final SAML2Configuration config = new SAML2Configuration();
        config.setHttpClient(httpClient.build());
        config.setKeystoreResourceUrl("http://localhost:8081/keystore");
        config.setServiceProviderMetadataResource(new FileUrlResource(new URL("http://localhost:8081/saml")));

		// Other Stuff...

        config.init();
        return config;
```

## Non-Goals

Note that the changes target the pac4j-saml module only. OpenSAML v3 is out of scope.

## TODO

- [X] Update docs after reviews
- [ ] Functional tests against CAS, etc.

### CODE REVIEW

File: pac4j-saml/src/main/java/org/pac4j/saml/metadata/SAML2HttpUrlMetadataGenerator.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+import org.opensaml.saml.metadata.resolver.impl.HTTPMetadataResolver;
+import org.springframework.core.io.Resource;
+
+import java.net.URL;
+import java.time.Duration;
+
+/**
+ * Generates metadata object with standard values and overridden user defined values.
+ *
+ * @author Michael Remond
```
Comment (leleuj) [inline pac4j-saml/src/main/java/org/pac4j/saml/metadata/SAML2HttpUrlMetadataGenerator.java:20]: This should be changed.

File: pac4j-saml/src/main/java/org/pac4j/saml/metadata/keystore/BaseSAML2KeystoreGenerator.java [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import java.security.cert.Certificate;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Date;
+
+/**
+ * This is {@link BaseSAML2KeystoreGenerator}.
+ *
+ * @author Misagh Moayyed
+ */
```
Comment (leleuj) [inline pac4j-saml/src/main/java/org/pac4j/saml/metadata/keystore/BaseSAML2KeystoreGenerator.java:40]: version

File: pac4j-saml/src/main/java/org/pac4j/saml/metadata/keystore/SAML2FileSystemKeystoreGenerator.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.regex.Pattern;
+
+/**
+ * This is {@link SAML2FileSystemKeystoreGenerator}.
+ *
+ * @author Misagh Moayyed
+ */
```
Comment (leleuj) [inline pac4j-saml/src/main/java/org/pac4j/saml/metadata/keystore/SAML2FileSystemKeystoreGenerator.java:25]: version

File: pac4j-saml/src/main/java/org/pac4j/saml/metadata/keystore/SAML2HttpUrlKeystoreGenerator.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.Base64;
+
+/**
+ * This is {@link SAML2HttpUrlKeystoreGenerator}.
+ *
+ * @author Misagh Moayyed
+ */
```
Comment (leleuj) [inline pac4j-saml/src/main/java/org/pac4j/saml/metadata/keystore/SAML2HttpUrlKeystoreGenerator.java:27]: version

**Comment by leleuj** [comment]
This looks good to me.

As it's a big change, I thought it should ship with 4.1.0 (September), but in fact, as it breaks no API, we can do that in version 4.0.1.

Please add some documentation: https://github.com/pac4j/pac4j/blob/master/documentation/docs/clients/saml.md and update the Javadoc.

Then, I'll merge.

**Comment by mmoayyed** [comment]
Sounds good, thanks for the review. I would like to borrow some time to test the changes with cas so I know everything works correctly. It will take a few days and I'm hoping to finalize by Friday.

**Comment by leleuj** [comment]
No problem. I don't plan to release the version 4.0.1 before next week or the following week.

**Comment by leleuj** [comment]
@mmoayyed One more thing: can you drop a line in the release notes as well: https://github.com/pac4j/pac4j/blob/master/documentation/docs/release-notes.md? Thx

**Comment by mmoayyed** [comment]
Found one breaking change and posted PR: https://github.com/pac4j/pac4j/pull/1585

**Comment by mmoayyed** [comment]
This should be ready. Let's get this merged, and I will resume testing more. So far, it's working as expected.

**Comment by leleuj** [comment]
OK. Waiting for the build to pass before merging...

**Comment by mmoayyed** [comment]
PS This feature is mainly going to be used and tested with Apache Syncope: https://github.com/apache/syncope/pull/181

**Comment by leleuj** [comment]
Excellent! Nothing better than a feature backed by real use cases.

**Comment by leleuj** [comment]
Merged!

**Comment by mmoayyed** [comment]
Thanks much!

**Pull request merged**